### PR TITLE
Even more resilient version.txt

### PR DESCRIPTION
**Pull request by wpedrak**
Follow up on https://github.com/kubernetes/kubernetes/pull/106653

#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:
On top of https://github.com/kubernetes/kubernetes/pull/106653 it helps to prevent etcd crash-loop in case of disk space shortage (see https://github.com/kubernetes/kubernetes/issues/98989)

#### Special notes for your reviewer:

This is complementary PR to https://github.com/kubernetes/kubernetes/pull/106653 created as suggested in  https://github.com/kubernetes/kubernetes/pull/106653#discussion_r756087078

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: cluster/images/etcd/migrate/data_dir.go [hunk 1]
```diff
@@ -140,10 +144,40 @@ func (v *VersionFile) Read() (*EtcdVersionPair, error) {
...
+		return nil
+	}
+	// We do write + rename instead of just write to protect from version.txt
+	// corruption under full disk condition.
+	// See https://github.com/kubernetes/kubernetes/issues/98989.
+	err = ioutil.WriteFile(v.nextPath(), []byte(vp.String()), 0666)
+	if err != nil {
+		return fmt.Errorf("failed to write new version file %s: %v", v.nextPath(), err)
+	}
+	return os.Rename(v.nextPath(), v.path)
```
Comment (wojtek-t) [inline cluster/images/etcd/migrate/data_dir.go:180]: @wpedrak - can you please rebase this PR after the previous one merged?

Comment (wpedrak) [reply cluster/images/etcd/migrate/data_dir.go]: done

**Comment by mm4tt** [comment]
/lgtm
/approve

/hold
Waiting for @wojtek-t to take a look as well

File: cluster/images/etcd/migrate/data_dir.go [hunk 2]
```diff
@@ -144,12 +144,36 @@ func (v *VersionFile) Read() (*EtcdVersionPair, error) {
 	return vp, nil
 }
 
+// contains returns true iff VersionFile exists and contains given EtcdVersionPair.
+func (v *VersionFile) contains(vp *EtcdVersionPair) (bool, error) {
```
Comment (wojtek-t) [inline cluster/images/etcd/migrate/data_dir.go:148]: I would actually call this function "equals" - effectively you check if its content equals to the given vp.

**Comment by wojtek-t** [comment]
One nit - though approving. Please hold cancel when you apply it.

/approve

File: cluster/images/etcd/migrate/data_dir.go (hunk 2 above)
Comment (wpedrak) [reply cluster/images/etcd/migrate/data_dir.go:148]: I'd argue that I'm checking if this file *contains* given vp. I'd expect `equals` method to take another `VersionFile` and compare them somehow. With that said, I don't have strong opinion on that.

Comment (wojtek-t) [reply cluster/images/etcd/migrate/data_dir.go:148]: I prefer equals, but I'm not going to fight until death for it.

Comment (wpedrak) [reply cluster/images/etcd/migrate/data_dir.go]: changed to `equals`

**Comment by wojtek-t** [comment]
/lgtm
/approve

/hold cancel

Thanks!

**Pull request merged**